### PR TITLE
feat(gossipsub): introduce backpressure

### PR DESCRIPTION
**Pull request by jxs**
## Description

superseeds #4914 with some changes and improvements, namely:

- introduce a `Delay` for `Forward` and `Publish` messages, messages that take more than the configured delay to be sent are discarded
- introduce scoring and penalize slow peers
- remove control pool
- report slow peers with the number of failed messages

### CODE REVIEW

File: protocols/gossipsub/src/behaviour.rs
Comment (ackintosh) [inline protocols/gossipsub/src/behaviour.rs]:
```suggestion
            for sender_queue in self.connected_peers.values().map(|v| &v.sender) {
```

File: protocols/gossipsub/src/peer_score.rs
Comment (ackintosh) [inline protocols/gossipsub/src/peer_score.rs]: We should move this nested `if` statement (`Slow peer weighting`) outside of `P7: behavioural pattern penalty`?

https://github.com/sigp/lighthouse/blob/bbe9242811b5dc14d991506bc4d076f0ce98b3cb/beacon_node/lighthouse_network/gossipsub/src/peer_score.rs#L339-L350

**Review by ackintosh — COMMENTED** [review]
I’ve left a few small comments.

File: protocols/gossipsub/src/peer_score/params.rs
Comment (guillaumemichel) [inline protocols/gossipsub/src/peer_score/params.rs:177]: where do these numbers come from?

File: protocols/gossipsub/src/config.rs
Comment (guillaumemichel) [inline protocols/gossipsub/src/config.rs:810]: this setter has the same name as the getter. generally the `with_` prefix is used for setters in the code base
```suggestion
    pub fn with_connection_handler_queue_len(&mut self, len: usize) -> &mut Self {
```

File: protocols/gossipsub/src/config.rs
Comment (guillaumemichel) [inline protocols/gossipsub/src/config.rs:817]:
```suggestion
    pub fn with_publish_queue_duration(&mut self, duration: Duration) -> &mut Self {
```

File: protocols/gossipsub/src/config.rs
Comment (guillaumemichel) [inline protocols/gossipsub/src/config.rs:824]:
```suggestion
    pub fn with_forward_queue_duration(&mut self, duration: Duration) -> &mut Self {
```

File: protocols/gossipsub/src/types.rs
Comment (guillaumemichel) [inline protocols/gossipsub/src/types.rs]: what is the reasoning of using `cap / 2` instead of `cap`?

File: protocols/gossipsub/src/types.rs
Comment (guillaumemichel) [inline protocols/gossipsub/src/types.rs]: is there a way to bound this channel?

File: protocols/gossipsub/src/types.rs
Comment (guillaumemichel) [inline protocols/gossipsub/src/types.rs]:
```suggestion
    pub(crate) fn priority_queue_len(&self) -> usize {
```

File: protocols/gossipsub/src/types.rs
Comment (guillaumemichel) [inline protocols/gossipsub/src/types.rs]:
```suggestion
    pub(crate) fn non_priority_queue_len(&self) -> usize {
```

File: protocols/gossipsub/src/types.rs
Comment (guillaumemichel) [inline protocols/gossipsub/src/types.rs]: what is a reasonable value for `cap`? should `RpcSender` also implement `Default` and provide a safe default cap?

File: protocols/gossipsub/src/types.rs
Comment (guillaumemichel) [inline protocols/gossipsub/src/types.rs]: `self.cap` is only used in `publish`. why?

File: protocols/gossipsub/src/types.rs
Comment (guillaumemichel) [inline protocols/gossipsub/src/types.rs]: this is substracting from `self.priority_len`. does it sometime also increase?

File: protocols/gossipsub/src/peer_score.rs
Comment (guillaumemichel) [inline protocols/gossipsub/src/peer_score.rs:444]: the default value of `slow_peer_decay` is set to `0.2`, which means that every heartbeat the `slow_peer_penalty` is divided by `5`. this seems like a lot, is the logic documented somewhere?

File: protocols/gossipsub/src/behaviour.rs
Comment (guillaumemichel) [inline protocols/gossipsub/src/behaviour.rs]: what is the reasoning for removing these functions? keeping functions shorts generally improves readability

**Review by guillaumemichel — COMMENTED** [review]
It would be great to add new tests for the introduced backpressure

File: protocols/gossipsub/src/behaviour.rs
Comment (jxs) [reply protocols/gossipsub/src/behaviour.rs]: thanks Akihito, updated!

File: protocols/gossipsub/src/peer_score.rs
Comment (jxs) [reply protocols/gossipsub/src/peer_score.rs]: right, thanks Akihito. Updated!

File: protocols/gossipsub/src/behaviour.rs
Comment (jxs) [reply protocols/gossipsub/src/behaviour.rs]: the control pool was removed in the process, we rather just try to send upfront and deal with the latency with the backpressure mechanism. `send_message` sent the messages to the `ConnectionHandler` via `Swarm`. 
This has been replaced by the more specific message types in `RpcSender` that send directly using channels to the `ConnectionHandler`

File: protocols/gossipsub/src/config.rs
Comment (jxs) [reply protocols/gossipsub/src/config.rs:810]: yeah, I am aware the thing is that `gossipsub` doesn't follow that logic, I'd suggest we do that for all the methods on another PR wdyt?

File: protocols/gossipsub/src/types.rs
Comment (jxs) [reply protocols/gossipsub/src/types.rs]: `RpcSender` and `RpcSender::new` are `pub(crate)`, it's cap is specified on `config.rs`:
```
/// The max number of messages a `ConnectionHandler` can buffer. The default is 5000.
    pub fn connection_handler_queue_len(&self) -> usize {
        self.connection_handler_queue_len
    }
```

File: protocols/gossipsub/src/types.rs
Comment (jxs) [reply protocols/gossipsub/src/types.rs]: we don't want to bound the channel, cause we still need to be able to send control messages like `GRAFT`, `PRUNE``, `SUBSCRIBE` and `UNSUBSCRIBE`, that's why we define `cap` as the capacity.
 But thanks for the question Gui left a comment explaining this!

File: protocols/gossipsub/src/types.rs
Comment (jxs) [reply protocols/gossipsub/src/types.rs]: cause we have two channels `priority` and `non_priority` left a comment elaborating on that, thanks Gui

File: protocols/gossipsub/src/types.rs
Comment (jxs) [reply protocols/gossipsub/src/types.rs]: because `publish` calls to `self.priority_sender` which is unbounded, and we need it unbounded to send control messages such as `GRAFT`, `PRUNE`, `SUBSCRIBE`, and `UNSUBSCRIBE`. Left a comment explaining this, thanks Gui!

File: protocols/gossipsub/src/types.rs
Comment (jxs) [reply protocols/gossipsub/src/types.rs]: thanks Gui, makes sense

File: protocols/gossipsub/src/types.rs
Comment (jxs) [reply protocols/gossipsub/src/types.rs]: thanks!

File: protocols/gossipsub/src/types.rs
Comment (jxs) [reply protocols/gossipsub/src/types.rs]: yeah, on line `697` everytime we send a publish message, here we are decreasing the counter as the message as been `unqueued`

**Review by jxs — COMMENTED** [review]
Thanks for the reviews Akihito and Gui!

File: protocols/gossipsub/src/types.rs
Comment (guillaumemichel) [reply protocols/gossipsub/src/types.rs]: So IIUC the `priority_sender` will not send `PUBLISH` messages if its `queue_len` is above `cap`, but it will send `GRAFT`, `PRUNE`, `SUBSCRIBE` and `UNSUBSCRIBE` anyway in an unbounded way?

File: protocols/gossipsub/src/types.rs
Comment (guillaumemichel) [reply protocols/gossipsub/src/types.rs]: alright, I had missed it

File: protocols/gossipsub/src/peer_score/params.rs
Comment (jxs) [reply protocols/gossipsub/src/peer_score/params.rs:177]: cc @AgeManning

File: protocols/gossipsub/src/peer_score.rs
Comment (jxs) [reply protocols/gossipsub/src/peer_score.rs:444]: cc @AgeManning

File: protocols/gossipsub/src/peer_score/params.rs
Comment (AgeManning) [reply protocols/gossipsub/src/peer_score/params.rs:177]: They are made up as suggestions. 

These are rough estimates from what we expect on real networks. I.e They are minor penalties relative to the behaviour penalty. 

We didn't want to penalize peers too much, so having 50 failures to 1 behaviour penalty seemed reasonable.

Note that these are not on by default and it is generally recommended that users set their own scoring based on their own networks.

File: protocols/gossipsub/src/peer_score.rs
Comment (AgeManning) [reply protocols/gossipsub/src/peer_score.rs:444]: We set the decay to be the same as the behaviour penalty decay. 

The score is lower, because we were expecting there to be more instances of this than the behaviour penalty.

The scoring mechanism itself is not enabled by default. When enabled, it is known to be highly sensitive to specific networks. Using default scoring typically does not work on arbitrary networks. 

Every use of gossipsub scoring I've seen on any real network has been custom configurable. I think these default values are not that important, and because these new scoring parameters are not in the specs, we set the default values of them to be small so that they would have minimal effect with the spec'd value scores. 

A rapid decay relative to a small score means the contributions from these new parameters are small.

File: protocols/gossipsub/src/peer_score/params.rs
Comment (guillaumemichel) [reply protocols/gossipsub/src/peer_score/params.rs:177]: > We didn't want to penalize peers too much, so having 50 failures to 1 behaviour penalty seemed reasonable.

OK, makes sense to me! Since this is defining the default for `PeerScoreParams`, it should be a safe and documented value (even if not enabled by default). IMO a comment saying that the `slow_peer_weight` is 50x lower than `behaviour_penalty_weight` would help.

Comment (jxs) [reply protocols/gossipsub/src/peer_score/params.rs:177]: updated, thanks Gui!

**Review by jxs — COMMENTED** [review]
> It would be great to add new tests for the introduced backpressure

yup for sure thanks for the suggestion Gui! Added some tests

File: protocols/gossipsub/src/types.rs
Comment (jxs) [reply protocols/gossipsub/src/types.rs]: yeah exactly!

**Review by ackintosh — APPROVED** [review]
Thanks João, LGTM!

File: protocols/gossipsub/src/types.rs
Comment (guillaumemichel) [reply protocols/gossipsub/src/types.rs]: Following the principle to [bound everything](https://github.com/libp2p/rust-libp2p/blob/master/docs/coding-guidelines.md#bound-everything): it would be great to bound the channel even for `GRAFT`, etc. messages. IIUC this would mean creating a new priority class. There would be 1. low priority, 2. normal priority (`PUBLISH`) and 3. high priority (`GRAFT`, etc.).

Not sure it would fit in the scope of this PR, but switching to using configurable priority queues, where the number of priority classes can be custom, could make sense.

Comment (jxs) [reply protocols/gossipsub/src/types.rs]: fair point yeah, we discussed this when implementing originally in our fork, I remember having similar opinion to yours Gui, i.e. if we can't send publish messages we also won't be able to send `GRAFT` `PRUNE` etc, but afair @AgeManning had a different opinion, do you rememebr why age?

File: protocols/gossipsub/src/behaviour.rs
Comment (elenaf9) [inline protocols/gossipsub/src/behaviour.rs]:
```suggestion
    /// Returns `true` if we were subscribed to this topic.
```

File: protocols/gossipsub/src/behaviour.rs
Comment (elenaf9) [inline protocols/gossipsub/src/behaviour.rs]: Nit: You could do an early return here to remove one layer of nesting:
```rs
if recipient_peers.is_empty() {
    return false
}
for peer_id in recipient_peers.iter() { ... }
tracing::debug!("Completed forwarding message");
true
```

File: protocols/gossipsub/src/types.rs
Comment (elenaf9) [inline protocols/gossipsub/src/types.rs]: Apologies if that has already been discussed, but why use `async_channel`, and not `futures::channel` like most other protocols?

File: protocols/gossipsub/src/types.rs
Comment (elenaf9) [inline protocols/gossipsub/src/types.rs]: Nit: I would expect a `types` module to not implement any major logic apart from converting and formatting.

Wdyt of moving the `RpcSender` and `RpcReceiver` into a separate module `behavior/rpc.rs` or similar?

File: protocols/gossipsub/src/types.rs
Comment (elenaf9) [inline protocols/gossipsub/src/types.rs]: I find the current description a bit confusing, Wdyt think of the below, based on `Config::{forward, publish}_queue_duration`

```suggestion
    /// Publish a Gossipsub message on network.`timeout` limits the duration the message 
    /// can wait to be sent before it is abandoned.
    Publish { message: RawMessage, timeout: Delay },
    /// Forward a Gossipsub message on network. `timeout` limits the duration the message 
    /// can wait to be sent before it is abandoned.
    Forward { message: RawMessage, timeout: Delay },
```

File: protocols/gossipsub/src/handler.rs
Comment (elenaf9) [inline protocols/gossipsub/src/handler.rs:413]: Drain all stale messages in the queue?
```suggestion
        // Drop stale messages in the queue.
        while let Poll::Ready(Some(rpc)) = self.send_queue.poll_stale(cx) {
            return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(
                HandlerEvent::MessageDropped(rpc),
            ));
        }
```

**Review by elenaf9 — COMMENTED** [review]
Few comments and nit picks.

File: protocols/gossipsub/src/types.rs
Comment (elenaf9) [inline protocols/gossipsub/src/types.rs]:
```suggestion
    /// The rpc sender to the connection handler(s).
```

File: protocols/gossipsub/src/types.rs
Comment (elenaf9) [inline protocols/gossipsub/src/types.rs:232]:
```suggestion
pub struct PeerInfo {
    pub peer_id: Option<PeerId>,
```
`PeerInfo` isn't re-exported from lib.rs, so I don't think this is needed?

File: protocols/gossipsub/src/types.rs
Comment (elenaf9) [inline protocols/gossipsub/src/types.rs]: Nit: first update the metrics, then send the message, so the message doesn't need to be cloned?

File: protocols/gossipsub/src/types.rs
Comment (elenaf9) [inline protocols/gossipsub/src/types.rs]:
```suggestion
    /// Send a `RpcOut::Forward` message to the `RpcReceiver`.
    /// This is low priority. If the queue is full the message is discarded and `Err` returned.
```

File: protocols/gossipsub/src/types.rs
Comment (elenaf9) [inline protocols/gossipsub/src/types.rs]:
```suggestion
    /// Capacity of the priority channel for `Publish` messages.
    cap: usize,
```
Also maybe renamed to `priority_cap`, to avoid confusion with the `cap` parameter in `RpcSender::new`?

File: protocols/gossipsub/src/behaviour.rs
Comment (elenaf9) [inline protocols/gossipsub/src/behaviour.rs]: Also add to metrics that the message was dropped?

File: protocols/gossipsub/src/behaviour.rs
Comment (elenaf9) [inline protocols/gossipsub/src/behaviour.rs]: Also add to metrics that publish message was dropped?

File: protocols/gossipsub/src/config.rs
Comment (elenaf9) [inline protocols/gossipsub/src/config.rs]:
```suggestion
                connection_handler_forward_duration: Duration::from_secs(1),
```

For consistency?

File: protocols/gossipsub/src/error.rs
Comment (elenaf9) [inline protocols/gossipsub/src/error.rs]:
```suggestion
    /// Messages could not be sent because the queues for all peers were full. The usize represents the
    /// number of peers that were attempted.
```

File: protocols/gossipsub/src/handler.rs
Comment (elenaf9) [inline protocols/gossipsub/src/handler.rs:273]: Would it make sense to move the logic for draining stale messages to the top? Then this check here wouldn't be needed anymore I think.

File: protocols/gossipsub/src/types.rs
Comment (elenaf9) [inline protocols/gossipsub/src/types.rs]: Isn't for example in case of a failed published message both `priority` **and** `publish` increased, and thus it would count twice into this sum?

File: protocols/gossipsub/src/behaviour.rs
Comment (elenaf9) [reply protocols/gossipsub/src/behaviour.rs]: Sorry for unresolving this but I was just wondering: couldn't we instead here in `send_message` simply call something like `RpcSender::send_message`?
`RpcSender::send_message` could then match on the `RpcOut` enum variant to decide which channel to send the message on.
The advantages would be:
1. smaller diff in the behavior code
2. could keep the control pool logic to bundle control messages
3. no separate `RpcSender::{graft(), prune(), publish(), ..}` etc methods needed anymore and thus code deduplication. The internal channels are type `RpcOut` anyway.

File: protocols/gossipsub/src/types.rs
Comment (jxs) [reply protocols/gossipsub/src/types.rs]: @elenaf9 found the previous discussion Gui: https://github.com/libp2p/rust-libp2p/issues/4667#issuecomment-1788658203:
>Have a queue for Subscriptions, Control and Publish. For now, this can be unbounded. At least memory footprint wise, Subscription and Control likely don't play a big role.

the pros of this is that if we can at least keep with network health, we'll do it.

Comment (guillaumemichel) [reply protocols/gossipsub/src/types.rs]: Alright, we can keep it like this, and we'll revisit later if the priority queue grows too large.

**Review by guillaumemichel — APPROVED** [review]

File: protocols/gossipsub/src/behaviour.rs
Comment (elenaf9) [inline protocols/gossipsub/src/behaviour.rs:2768]: Why was this logic moved to `handle_established_inbound_connection` and `handle_established_outbound_connection`?

File: protocols/gossipsub/src/types.rs
Comment (elenaf9) [reply protocols/gossipsub/src/types.rs:232]: Nvm, is is reaquired because of the "unreachable pub item" lint.

File: protocols/gossipsub/src/behaviour.rs
Comment (elenaf9) [reply protocols/gossipsub/src/behaviour.rs]: I played around a bit with it and drafted something in https://github.com/jxs/rust-libp2p/pull/12. But I am also fine with keeping the current implementation, so up to you @jxs :)

File: protocols/gossipsub/src/types.rs
Comment (elenaf9) [reply protocols/gossipsub/src/types.rs]: Discussed out of band with @jxs. The `async_channel` is needed because we need multiple consumers, and the `futures::channel` is single consumer only.

File: protocols/gossipsub/src/behaviour.rs
Comment (elenaf9) [reply protocols/gossipsub/src/behaviour.rs]: Obsolete with jxs#12.

File: protocols/gossipsub/src/types.rs
Comment (elenaf9) [reply protocols/gossipsub/src/types.rs]: Obsolete with jxs#12.

File: protocols/gossipsub/src/types.rs
Comment (elenaf9) [reply protocols/gossipsub/src/types.rs]: Obsolete with jxs#12.

File: protocols/gossipsub/src/behaviour.rs
Comment (elenaf9) [reply protocols/gossipsub/src/behaviour.rs]: jxs#12 merged.

File: protocols/gossipsub/src/types.rs
Comment (jxs) [reply protocols/gossipsub/src/types.rs]: yeah makes sense, thanks Elena!

File: protocols/gossipsub/src/types.rs
Comment (jxs) [reply protocols/gossipsub/src/types.rs]: thanks!

File: protocols/gossipsub/src/error.rs
Comment (jxs) [reply protocols/gossipsub/src/error.rs]: thanks!

File: protocols/gossipsub/src/config.rs
Comment (jxs) [reply protocols/gossipsub/src/config.rs]: yup thanks!

File: protocols/gossipsub/src/types.rs
Comment (jxs) [reply protocols/gossipsub/src/types.rs]: yup, that's right, thanks Elena. Updated to return `self.priority + self.non_priority`

File: protocols/gossipsub/src/handler.rs
Comment (jxs) [reply protocols/gossipsub/src/handler.rs:413]: I think it's the same no? Since whenever the condition matches the function will return, that's why I prefer `if`, `while` gives a wrong impression

File: protocols/gossipsub/src/behaviour.rs
Comment (jxs) [reply protocols/gossipsub/src/behaviour.rs:2768]: because we need to add the `ConnectionHandler` to the `PeerConnections` and we only have them on those methods

File: protocols/gossipsub/src/handler.rs
Comment (jxs) [reply protocols/gossipsub/src/handler.rs:273]: yeah makes sense, but can we leave it to the subsequent PR along with avoiding the iteration optimizations wdyt?

**Review by jxs — COMMENTED** [review]
Think I addressed everything, can you take another look Elena? Thanks!

File: protocols/gossipsub/src/types.rs
Comment (jxs) [reply protocols/gossipsub/src/types.rs]: yeah makes sense, did that and renamed them to just `Sender` and `Receiver` no need for the `Rpc` prefixes

File: protocols/gossipsub/src/handler.rs
Comment (elenaf9) [reply protocols/gossipsub/src/handler.rs:413]: Makes sense :+1:

File: protocols/gossipsub/src/behaviour.rs
Comment (elenaf9) [reply protocols/gossipsub/src/behaviour.rs:2768]: Ahh yes makes sense!

**Comment by elenaf9** [comment]
I think the new `rpc` module was accidentally not included in 602c46f? 😇
But everything else looks good to me! Just one remaining open comment about whether in the case that the channel is full and publishing failed, we should also update the metrics same as when a message is dropped?

**Comment by jxs** [comment]
> I think the new `rpc` module was accidentally not included in [602c46f](https://github.com/libp2p/rust-libp2p/commit/602c46f704df3b13754962dc9ac4b4666761e6d6)? 😇 But everything else looks good to me! Just one remaining open comment about whether in the case that the channel is full and publishing failed, we should also update the metrics same as when a message is dropped?

ah right! Added the `rpc` module. The unresolved conversations were from before your PR was merged, now metrics are being record on `send_message` and cover the cases you commented right?

**Review by elenaf9 — APPROVED** [review]
LGTM!

**Pull request merged**
